### ISSUE TITLE
HHC17512E under Ubuntu 18.04

### ISSUE THREAD

**Issue by Peter-J-Jansen** [SDL-Hercules-390/hyperion#168]
I have created ooRexx under Ubuntu 18.04 and MacOS. The first time I start SDL-Hyperion under Ubuntu 18.04 after a reboot, it generates HHC17512E; only the first time. An easy circumvention is to just start a simple REXX which does not have to do anything useful, prior to the SDL-Hyperion start. 

```
11:12:02 HHC00100I Thread id 00007f09771ab740, prio 5, name 'impl_thread' started
11:12:02 HHC00100I Thread id 00007f0974d7c700, prio 4, name 'logger_thread' started
11:12:02 HHC01413I Hercules version 4.2.0.0-SDL-g70d850c0-modified (4.2.0.0)
11:12:02 HHC01414I (C) Copyright 1999-2018 by Roger Bowler, Jan Jaeger, and others
11:12:02 HHC01417I ** The SoftDevLabs version of Hercules **
11:12:02 HHC01415I Build date: Dec 20 2018 at 17:09:17
11:12:02 HHC01417I Built with: GCC 7.3.0
11:12:02 HHC01417I Build type: GNU/Linux x86_64 host architecture build
11:12:02 HHC01417I Modes: S/370 ESA/390 z/Arch
11:12:02 HHC01417I Max CPU Engines: 16
11:12:02 HHC01417I Using   shared libraries
11:12:02 HHC01417I Using   setresuid() for setting privileges
11:12:02 HHC01417I Using   POSIX threads Threading Model
11:12:02 HHC01417I Using   Error-Checking Mutex Locking Model
11:12:02 HHC01417I With    Shared Devices support
11:12:02 HHC01417I With    Dynamic loading support
11:12:02 HHC01417I With    External GUI support
11:12:02 HHC01417I With    IPV6 support
11:12:02 HHC01417I With    HTTP Server support
11:12:02 HHC01417I With    sqrtl support
11:12:02 HHC01417I With    SIGABEND handler
11:12:02 HHC01417I With    CCKD BZIP2 support
11:12:02 HHC01417I With    HET BZIP2 support
11:12:02 HHC01417I With    ZLIB support
11:12:02 HHC01417I With    Regular Expressions support
11:12:02 HHC01417I With    Object REXX support
11:12:02 HHC01417I Without Regina REXX support
11:12:02 HHC01417I With    Automatic Operator support
11:12:02 HHC01417I Without National Language Support
11:12:02 HHC01417I With    CCKD64 Support
11:12:02 HHC01417I Machine dependent assists: cmpxchg1 cmpxchg4 cmpxchg8 hatomics=C11
11:12:02 HHC01417I Running on: pjvmubuntu (Linux-4.15.0-43-generic x86_64) MP=4
11:12:02 HHC01417I Built with decNumber external package version 3.68.0.79-g53f2512
11:12:02 HHC01417I Built with SoftFloat external package version 3.5.0.82-g1c66591
11:12:02 HHC01417I Built with telnet external package version 1.0.0.41-ged0ddec
11:12:02 HHC00018I Hercules is running in elevated mode
11:12:02 HHC00150I Crypto module loaded (C) Copyright 2003-2016 by Bernard van der Helm
11:12:02 HHC01417I Built with crypto external package version 1.0.0.26-gefe199e
11:12:02 HHC00151I Activated facility: Message Security Assist
11:12:02 HHC00151I Activated facility: Message Security Assist Extension 1, 2, 3 and 4
11:12:02 HHC17512E REXX(OORexx) RegisterHandlers: return code 0x000003EA from OORexxRegisterExit
11:12:02 HHC00007I Previous message from function 'RegisterRexxHandlersAndFunctions' at hRexxapi.c(1163)
11:12:02 HHC17511E REXX() Could not enable either Rexx package
11:12:02 HHC00007I Previous message from function 'Enable' at hRexx.c(510)
11:12:02 HHC17511E REXX() Could not enable default Rexx package
11:12:02 HHC00007I Previous message from function 'InitializeRexx' at hRexx.c(261)
11:12:02 HHC17500I REXX() Mode            : Command
11:12:02 HHC17500I REXX() MsgLevel        : Off
11:12:02 HHC17500I REXX() MsgPrefix       : Off
11:12:02 HHC17500I REXX() ErrPrefix       : Off
11:12:02 HHC17500I REXX() Resolver        : On
11:12:02 HHC17500I REXX() SysPath    ( 8) : On
11:12:02 HHC17500I REXX() RexxPath   ( 0) :
11:12:02 HHC17500I REXX() Extensions ( 8) : .REXX:.rexx:.REX:.rex:.CMD:.cmd:.RX:.rx
```

Hopefully this is not related to Issue # 141 :-)

This is a very low priority item.

Cheers,

Peter

**Comment by Fish-Git** [comment]
> I have created ooRexx under Ubuntu 18.04 and MacOS.

What does that mean? "I have created..."? Do you mean you built it from source yourself? And then installed it yourself?

> The first time I start SDL-Hyperion under Ubuntu 18.04 after a reboot, it generates HHC17512E; only the first time. An easy circumvention is to just start a simple REXX which does not have to do anything useful, prior to the SDL-Hyperion start.

I recall having a similar problem on Linux and while I don't recall all the details of what I had to do to fix it, I did find some web pages that might explain what's going on. (I believe the `rxapi` daemon is not being automatically started upon reboot.)

* [Hercules-REXX Integration](https://hercules-390.github.io/html/rexxinteg.html#oorexx)

* [Open Object Rexx Release Notes (PDF)](http://wi.wu.ac.at:8002/rgf/tmp/201202/readme.pdf) &nbsp; (*)

Page 21 of the Release Notes PDF says:

> Part of ooRexx is rxapi which is a daemon process that manages all data that can persist across interpreter invocations or is used for cross-process communications. The rxapi process manages the Rexx data queues, the macrospace, and all of the external function, subcommand handler and exit registrations.

Which means the `rxapi` daemon _must_ be started before Hercules starts (since Hercules uses subcommand handlers and exit registrations).

It then goes on to say:

> If rxapi is not installed as a service, the ﬁrst time the interpreter executes, and needs the function provided by rxapi, the intepreter will start the daemon process. Once started, the rxapi process will continue to run until the system is halted.

Which I believe explains the behavior you are seeing.

Steve's "other Hyperion" Rexx web page also mentions most of this too.

**Bottom line:** it doesn't sound like you have ooRexx installed properly. You need to do whatever it is to ensure the `rxapi` daemon (`rxapid`??) is always automatically started upon every reboot. Do that, and I'm sure this problem will go away.

&nbsp;
(*) The PDF is for version 4.1 but I believe the `rxapi` requirement is still true for _all_ versions of ooRexx since then (or possible since _before_ then).
